File: app/src/main/java/com/xyoye/dandanplay/ui/weight/anim/path/TextPathAnimView.java
package com.xyoye.dandanplay.ui.weight.anim.path;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.animation.ValueAnimator;
import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Path;
import android.graphics.PathMeasure;
import android.util.AttributeSet;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.LinearInterpolator;

import androidx.annotation.Nullable;

import com.xyoye.dandanplay.R;

/**
 * Created by xyoye on 2019/9/15.
 */

public class TextPathAnimView extends View {
    private static final String TAG = TextPathAnimView.class.getSimpleName();

    private final int mTextBgColor;
    private final int mTextFgColor;
    private final boolean mIsLoop;
    private final int mAnimDuration;

    private Paint mPaint;
    private Path mAnimPath;
    private Path mSourcePath;
    private TextPath mSourceTextPath;
    private PathMeasure mPathMeasure;
    private ValueAnimator mAnimator;
    private AnimListener animListener;

    private int mPaddingLeft, mPaddingTop;

    public TextPathAnimView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public TextPathAnimView(Context context, @Nullable AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);

        TypedArray typedArray = context.obtainStyledAttributes(attrs, R.styleable.TextPathAnimView, defStyle, 0);
        mAnimDuration = typedArray.getInt(R.styleable.TextPathAnimView_duration, 1500);
        mTextBgColor = typedArray.getColor(R.styleable.TextPathAnimView_text_bg_color, Color.BLACK);
        mTextFgColor = typedArray.getColor(R.styleable.TextPathAnimView_text_fg_color, Color.WHITE);
        mIsLoop = typedArray.getBoolean(R.styleable.TextPathAnimView_loop, true);

        String contentText = typedArray.getString(R.styleable.TextPathAnimView_text);
        float sizeScale = typedArray.getFloat(R.styleable.TextPathAnimView_text_size_scale, dp2px(14));
        float stokeWidth = typedArray.getFloat(R.styleable.TextPathAnimView_text_stoke_width, 3f);
        float textInterval = typedArray.getDimension(R.styleable.TextPathAnimView_text_interval, dp2px(5));
        typedArray.recycle();

        mSourceTextPath = new TextPath(contentText, sizeScale, textInterval);
        mSourcePath = mSourceTextPath.getPath();

        mPaint = new Paint();
        mPaint.setAntiAlias(true);
        mPaint.setStyle(Paint.Style.STROKE);
        mPaint.setStrokeWidth(stokeWidth);

        mAnimPath = new Path();
        mPathMeasure = new PathMeasure();
    }

    private void initAnim() {
        mAnimator = ValueAnimator.ofFloat(0.0f, 1.0f);
        mAnimator.setInterpolator(new LinearInterpolator());
        mAnimator.setRepeatCount(ValueAnimator.INFINITE);
        mAnimator.addUpdateListener(animation -> {
            float stopD = mPathMeasure.getLength() * (float) animation.getAnimatedValue();
            mPathMeasure.getSegment(0, stopD, mAnimPath, true);
            invalidate();
        });

        mAnimator.addListener(new AnimatorListenerAdapter() {
            @Override
            public void onAnimationRepeat(Animator animation) {
                mPathMeasure.getSegment(0, mPathMeasure.getLength(), mAnimPath, true);
                mPathMeasure.nextContour();
                if (mPathMeasure.getLength() == 0) {
                    if (mIsLoop) {
                        mAnimPath.reset();
                        mAnimPath.lineTo(0, 0);
                        mPathMeasure.setPath(mSourcePath, false);
                        if (animListener != null) {
                            animListener.onLoop();
                        }
                    } else {
                        animation.end();
                    }
                }
            }

            @Override
            public void onAnimationStart(Animator animation) {
                super.onAnimationStart(animation);
                if (animListener != null) {
                    animListener.onStart();
                }
            }

            @Override
            public void onAnimationEnd(Animator animation) {
                super.onAnimationEnd(animation);
                if (animListener != null && !mIsLoop) {
                    animListener.onEnd();
                }
            }
        });
    }

    private float dp2px(float dpValue) {
        final float scale = getContext().getResources().getDisplayMetrics().density;
        return (int) (dpValue * scale + 0.5f);
    }

    @Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        super.onSizeChanged(w, h, oldw, oldh);
        mPaddingLeft = getPaddingLeft();
        mPaddingTop = getPaddingTop();
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
        int originWith = MeasureSpec.getSize(widthMeasureSpec);
        int originHeight = MeasureSpec.getSize(heightMeasureSpec);

        int newWidth = (int) mSourceTextPath.getWidth();
        int newHeight = (int) mSourceTextPath.getHeight();

        if (getLayoutParams().width == ViewGroup.LayoutParams.WRAP_CONTENT
                && getLayoutParams().height == ViewGroup.LayoutParams.WRAP_CONTENT) {
            setMeasuredDimension(newWidth, newHeight);
        } else if (getLayoutParams().width == ViewGroup.LayoutParams.WRAP_CONTENT) {
            setMeasuredDimension(newWidth, originHeight);
        } else if (getLayoutParams().height == ViewGroup.LayoutParams.WRAP_CONTENT) {
            setMeasuredDimension(originWith, newHeight);
        }
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);

        canvas.translate(mPaddingLeft, mPaddingTop);

        mPaint.setColor(mTextBgColor);
        canvas.drawPath(mSourcePath, mPaint);

        mPaint.setColor(mTextFgColor);
        canvas.drawPath(mAnimPath, mPaint);
    }

    public void startAnim() {
        if (null == mAnimator) {
            initAnim();
        } else if (mAnimator.isRunning()) {
            return;
        }

        mAnimPath.reset();
        mAnimPath.lineTo(0, 0);
        mPathMeasure.setPath(mSourcePath, false);
        int count = 0;
        while (mPathMeasure.getLength() != 0) {
            mPathMeasure.nextContour();
            count++;
        }
        mPathMeasure.setPath(mSourcePath, false);

        mAnimator.setDuration(mAnimDuration / count);
        mAnimator.start();
    }

    public void stopAnim() {
        if (null != mAnimator && mAnimator.isRunning()) {
            mAnimator.end();
        }
    }

    public void setAnimListener(AnimListener listener) {
        this.animListener = listener;
    }

    public interface AnimListener {
        void onStart();

        void onEnd();

        void onLoop();
    }
}

File: app/src/main/java/com/xyoye/dandanplay/utils/image_anim/path/TextPathAnimView.kt
package com.xyoye.dandanplay.utils.image_anim.path

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.animation.ValueAnimator
import android.content.Context
import android.graphics.*
import android.util.AttributeSet
import android.view.View
import android.view.ViewGroup
import android.view.animation.LinearInterpolator
import com.xyoye.dandanplay.R

/**
 * Created by xyoye on 2019/9/15.
 */

class TextPathAnimView : View {
    private val mTextBgColor: Int
    private val mTextFgColor: Int
    private val mIsLoop: Boolean
    private val mAnimDuration: Int
    private val mSourcePath: Path
    private val mSourceTextPath: TextPath

    private var mPaddingLeft = 0
    private var mPaddingTop = 0

    private val mPaint = Paint()
    private val mAnimPath = Path()
    private val mPathMeasure = PathMeasure()
    private val mAnimator = ValueAnimator.ofFloat(0.0f, 1.0f)

    private var animListener: AnimListener? = null

    constructor(context: Context) : this(context, null)

    constructor(context: Context, attrs: AttributeSet?) : this(context, attrs, 0)

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(
        context,
        attrs,
        defStyleAttr
    ) {
        mPaint.run {
            isAntiAlias = true
            style = Paint.Style.STROKE
        }

        context.obtainStyledAttributes(
            attrs,
            R.styleable.TextPathAnimView
        ).run {
            mAnimDuration = getInt(R.styleable.TextPathAnimView_duration, 1500)
            mTextBgColor = getColor(R.styleable.TextPathAnimView_text_bg_color, Color.BLACK)
            mTextFgColor = getColor(R.styleable.TextPathAnimView_text_fg_color, Color.WHITE)
            mIsLoop = getBoolean(R.styleable.TextPathAnimView_loop, true)
            val contentText = getString(R.styleable.TextPathAnimView_text) ?: ""
            val sizeScale = getFloat(R.styleable.TextPathAnimView_text_size_scale, dp2px(14f))
            val stokeWidth = getFloat(R.styleable.TextPathAnimView_text_stoke_width, 3f)
            val textInterval = getDimension(R.styleable.TextPathAnimView_text_interval, dp2px(5f))

            mSourceTextPath = TextPath(contentText, sizeScale, textInterval)
            mSourcePath = mSourceTextPath.path
            mPaint.strokeWidth = stokeWidth
            recycle()
        }

        initAnim()
    }

    private fun initAnim() {
        mAnimator.run {
            interpolator = LinearInterpolator()
            repeatCount = ValueAnimator.INFINITE

            addUpdateListener { animation: ValueAnimator ->
                val stopD = mPathMeasure.length * animation.animatedValue as Float
                mPathMeasure.getSegment(0f, stopD, mAnimPath, true)
                invalidate()
            }

            addListener(object : AnimatorListenerAdapter() {
                override fun onAnimationRepeat(animation: Animator) {
                    mPathMeasure.getSegment(
                        0f,
                        mPathMeasure.length,
                        mAnimPath,
                        true
                    )
                    mPathMeasure.nextContour()
                    if (mPathMeasure.length == 0f) {
                        if (mIsLoop) {
                            mAnimPath.reset()
                            mAnimPath.lineTo(0f, 0f)
                            mPathMeasure.setPath(mSourcePath, false)
                            animListener?.onLoop()
                        } else {
                            animation.end()
                        }
                    }
                }

                override fun onAnimationStart(animation: Animator) {
                    super.onAnimationStart(animation)
                    animListener?.onStart()
                }

                override fun onAnimationEnd(animation: Animator) {
                    super.onAnimationEnd(animation)
                    if (!mIsLoop) {
                        animListener?.onEnd()
                    }
                }
            })
        }
    }

    private fun dp2px(dpValue: Float): Float {
        val scale = context.resources.displayMetrics.density
        return dpValue * scale + 0.5f
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)
        mPaddingLeft = paddingLeft
        mPaddingTop = paddingTop
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        val originWith = MeasureSpec.getSize(widthMeasureSpec)
        val originHeight = MeasureSpec.getSize(heightMeasureSpec)
        val newWidth = mSourceTextPath.width.toInt()
        val newHeight = mSourceTextPath.height.toInt()
        if (layoutParams.width == ViewGroup.LayoutParams.WRAP_CONTENT
            && layoutParams.height == ViewGroup.LayoutParams.WRAP_CONTENT)
        {
            setMeasuredDimension(newWidth, newHeight)
        } else if (layoutParams.width == ViewGroup.LayoutParams.WRAP_CONTENT) {
            setMeasuredDimension(newWidth, originHeight)
        } else if (layoutParams.height == ViewGroup.LayoutParams.WRAP_CONTENT) {
            setMeasuredDimension(originWith, newHeight)
        }
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        canvas.translate(mPaddingLeft.toFloat(), mPaddingTop.toFloat())
        mPaint.color = mTextBgColor
        canvas.drawPath(mSourcePath, mPaint)
        mPaint.color = mTextFgColor
        canvas.drawPath(mAnimPath, mPaint)
    }

    fun startAnim() {
        if (mAnimator.isRunning) {
            return
        }
        mAnimPath.reset()
        mAnimPath.lineTo(0f, 0f)
        mPathMeasure.setPath(mSourcePath, false)
        var count = 0
        while (mPathMeasure.length != 0f) {
            mPathMeasure.nextContour()
            count++
        }
        mPathMeasure.setPath(mSourcePath, false)
        mAnimator.duration = mAnimDuration / count.toLong()
        mAnimator.start()
    }

    fun stopAnim() {
        if (mAnimator.isRunning) {
            mAnimator.end()
        }
    }

    fun setAnimListener(listener: AnimListener?) {
        animListener = listener
    }

    interface AnimListener {
        fun onStart()
        fun onEnd()
        fun onLoop()
    }
}